### PR TITLE
Make `SparseArray` store constructors

### PR DESCRIPTION
**Pull request by C0Florent**
## Changes made

This PR improves `SparseArray` by making it store a component constructor as a private member. This allows the addition of a new method: `SparseArray::constructAt` which constructs a new Component instance at a given index, using the stored constructor function.

This PR also changes SparseArray to use an `std::unordered_map<std::size_t, std::unique_ptr<Component>>` instead of a `std::vector<std::unique_ptr<Component>>` for storage. This should provide a better performance balance between memory usage and CPU performance (the vector alternative is faster but consumes more memory).

## Why?

The new `constructAt` method will help the R-Type client build components requested by the server more easily.

### CODE REVIEW

File: src/ECS/utilities/ValueIterator.hpp [hunk 1]
```diff
@@ -0,0 +1,25 @@
+/*
+** EPITECH PROJECT, 2024
+** BsRtype
```
Comment (Tugduoff) [inline src/ECS/utilities/ValueIterator.hpp:3]: Its RType not BsRtype ^^

File: src/ECS/utilities/SparseArray.hpp [hunk 2]
```diff
@@ -8,28 +8,41 @@
 #ifndef SPARSE_ARRAY_HPP
```
Comment (Tugduoff) [inline src/ECS/utilities/SparseArray.hpp:8]: The header contains a wrong note, its RType not BSRtype (it's my bad, but it could be time to change it !)

**Review by Tugduoff — COMMENTED** [review]
Except a small typo, everything looks good and functional. I already have infos on how it works since you told me earlier. Great step

File: src/ECS/utilities/ValueIterator.hpp [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+ */
+template<typename It>
+struct ValueIterator : public It {
+    ValueIterator(It src) : It(std::move(src)) {}
+
+    auto &operator*() const 
+    {
+        return It::operator*().second;
+    }
+};
```
Comment (ClementBaziret) [inline src/ECS/utilities/ValueIterator.hpp:25]: EOF newline

**Review by ClementBaziret — APPROVED** [review]
Seems good 🫡

File: src/ECS/utilities/ValueIterator.hpp (hunk 1 above)
Comment (Tugduoff) [reply src/ECS/utilities/ValueIterator.hpp]: its been resolved

File: src/ECS/utilities/SparseArray.hpp (hunk 2 above)
Comment (Tugduoff) [reply src/ECS/utilities/SparseArray.hpp:8]: It's been resolved as well

**Review by Tugduoff — DISMISSED** [review]

**Review by Tugduoff — APPROVED** [review]

**Review by tibaudlemaout — APPROVED** [review]
All of it seems good to me and with the explanation you gave me in person i have no questions. :+1:

**Pull request merged**